### PR TITLE
Highlight drop targets for palette items

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: elements/designer-stage/designer-stage.html [hunk 1]
```diff
@@ -405,8 +410,8 @@
         // TODO(justinfagnani): we might want to wait until we know we don't
         // have a double-click to initiate any action, but the redundant
         // selection that happens here sometimes is ok for now
-        var bounds = this.getBoundingClientRect();
-        this.client.selectElementAtPoint(e.clientX, e.clientY, bounds)
+        let position = domUtils.toLocalPosition(this, e.clientX, e.clientY);
+        this.client.selectElementAtPoint(position.x, position.y)
```
Comment (cdata) [inline elements/designer-stage/designer-stage.html]: This looks like a nice improvement :100:

File: elements/designer-stage/designer-stage.html [hunk 2]
```diff
@@ -431,32 +448,33 @@
               this.insertPosition));
         }
         this._hideHover();
+        this.insertPosition = null;
```
Comment (cdata) [inline elements/designer-stage/designer-stage.html]: Should this be a private property?

File: elements/designer-stage/designer-stage.html [hunk 3]
```diff
@@ -431,32 +448,33 @@
...
-        this.client.selectElementAtPoint(e.clientX, e.clientY, bounds)
-            .then(function(response) {
-              this._updateCurrentElement(response.elementInfo);
-              this._updateSelectionBounds(response.bounds);
-              this.enterTextMode();
-            }.bind(this));
+        let position = domUtils.toLocalPosition(this, e.clientX, e.clientY);
+        this.client.selectElementAtPoint(position.x, position.y)
+          .then(function(response) {
+            this._updateCurrentElement(response.elementInfo);
```
Comment (cdata) [inline elements/designer-stage/designer-stage.html]: This is orthogonal to this change, but `set` seems like a more appropriate verb than `update` to me. No action required, just observing.

File: src/dom-utils/dom-utils.js [hunk 4]
```diff
@@ -55,11 +55,31 @@ define('polymer-designer/dom-utils', function() {
...
+   *
+   * @param node {Node}
+   * @param clientX {number}
+   * @param clientY {number}
+   * @return {{
+   *   x: number,
+   *   y: number,
+   * }}
+   */
+  function toLocalPosition(node, clientX, clientY) {
```
Comment (cdata) [inline src/dom-utils/dom-utils.js]: me gusta :smile_cat:

File: src/protocol/DocumentClient.js [hunk 5]
```diff
@@ -61,6 +63,25 @@ define('polymer-designer/protocol/DocumentClient', function() {
...
+      return this.connection.request({
+        messageType: 'getElementsAtPoint',
+        x: x,
+        y: y,
+      }).then(function(response) {
+        let elements = response.elements;
+        for (let i in elements) {
+          let element = elements[i];
+          let id = element.elementInfo.id;
+          this.nodes.set(id, element);
```
Comment (cdata) [inline src/protocol/DocumentClient.js]: Interesting. Can you explain what this iteration over `elements` is doing? I have some guesses, but I would love to hear more.

File: src/protocol/DocumentServer.js [hunk 6]
```diff
@@ -89,25 +90,19 @@ define('polymer-designer/protocol/DocumentServer', [
...
-          // now at a top-level host, return the candidate
-          if (lastLightCandidate == null) {
-            lastLightCandidate = el;
-          }
-          break;
-        } else {
-          lastLightCandidate = el;
-        }
-        el = el.parentNode;
+    getElementsAtPoint(request) {
```
Comment (cdata) [inline src/protocol/DocumentServer.js]: Maybe add some docs? :scroll:

**Comment by cdata** [comment]
Minor questions, otherwise LGTM :+1:

File: elements/designer-stage/designer-stage.html (hunk 2 above)
Comment (justinfagnani) [inline elements/designer-stage/designer-stage.html]: yes

File: elements/designer-stage/designer-stage.html (hunk 3 above)
Comment (justinfagnani) [inline elements/designer-stage/designer-stage.html]: ack

File: src/protocol/DocumentClient.js (hunk 5 above)
Comment (justinfagnani) [inline src/protocol/DocumentClient.js]: It's registering node ids that we might not have seen yet. The client only knows about nodes that have been passed through the protocol before.

File: src/protocol/DocumentServer.js (hunk 6 above)
Comment (justinfagnani) [inline src/protocol/DocumentServer.js]: done
